gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -29,41 +29,188 @@
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
 import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.io.IOException;
 import java.util.List;
+import java.util.Vector;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.threeten.bp.Duration;
 
 /**
  * A {@link ManagedChannel} that will send requests round robin via a set of channels.
  *
  * <p>Package-private for internal use.
  */
 class ChannelPool extends ManagedChannel {
-  private final ImmutableList<ManagedChannel> channels;
+  private static final Logger LOG = Logger.getLogger(ChannelPool.class.getName());
+  // refresh every 55 minutes
+  private static Duration refreshPeriod = Duration.ofMinutes(55);
+  // maximum seconds to wait for old channel to terminate
+  static Duration terminationWait = Duration.ofMinutes(1);
+
+  private final List<ManagedChannel> channels;
+  private final List<AtomicInteger> requestCounters;
   private final AtomicInteger indexTicker = new AtomicInteger();
+  private final List<ReadWriteLock> locks;
   private final String authority;
+  private final boolean isRefreshing;
 
   /**
    * Initializes the channel pool. Assumes that all channels have the same authority.
    *
-   * @param channels a List of channels to pool.
+   * @param poolSize number of channels in the pool.
+   * @param channelFactory method to create the channels.
+   * @param executorService if set, schedule periodically refresh the channels.
    */
-  ChannelPool(List<ManagedChannel> channels) {
-    this.channels = ImmutableList.copyOf(channels);
+  ChannelPool(
+      int poolSize, final ChannelFactory channelFactory, ScheduledExecutorService executorService)
+      throws IOException {
+    channels = new Vector<>(poolSize);
+    requestCounters = new Vector<>(poolSize);
+    locks = new Vector<>(poolSize);
+    for (int i = 0; i < poolSize; i++) {
+      channels.add(channelFactory.createSingleChannel());
+      requestCounters.add(new AtomicInteger(0));
+      locks.add(new ReentrantReadWriteLock());
+    }
+    // if executorService is available, schedule channels to be refreshed
+    isRefreshing = executorService != null;
+
+    if (isRefreshing) {
+      // stagger the refresh to not overload GFE
+      // spread out the refresh between channels
+      Duration refreshDelay = refreshPeriod.dividedBy(poolSize);
+
+      // runnable inner class to refresh a channel in the list of channels
+      class RefreshSingleChannel implements Runnable {
+        private final int index;
+        private final List<ManagedChannel> channels;
```

Comment:
Please use AtomicReferenceArray instead. You have different threads writing and reading the elements. The changes one thread makes might no be visible to another thread